<table align="left">
  <td>
    <a href="https://colab.research.google.com/github/marco-canas/didactica_ciencia_datos/blob/main/propuesta/unidad_didac/1_statistic/1_Basic_Descriptive_Statistics/1_medidas_tendencia_central/1_media.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>
  </td>
</table>

## [Video de apoyo a la lectura interactiva y experimental de este cuaderno](https://www.youtube.com/watch?v=V9ikhkgm8ag)

## [Segundo video de apoyo a la lectura interactiva de este cuaderno](https://www.youtube.com/watch?v=2RXc920iTus)

# Medidas de tendencia central - La media

El modelo pedagógico que sustenta esta propuesta de Unidad didáctica para la enseñanza de la ciencia de datos es el de [Flipped Learning](https://www3.gobiernodecanarias.org/medusa/ecoescuela/pedagotic/files/2018/11/flipped-classroom.pdf) o de aprendizaje investido. 

### Metodología para la clase sobre la media como estadístico importante para un científico de datos

1. exploración de presaberes: listas, tuples, tipos de datos, acceso a los datos. 
2. Definir una función de usuario de Python que calcule la media de un conjunto de datos.
3. determinar la media con la librería NumPy a interpretarla.
4. Determinar la media con la librería scipy
5. Determinar la media con el módulo statistics
6. Realizar la tarea de fijación de conceptos y adquisición de habilidades
7. Realizar la evaluación de conceptos ganados y habilidades adquiridas.
8. Hacer la autoevaluación.

## Elementos previos o conceptos previos. 

* Listas y tuplas de python
* Funciones incorporadas de python
* Funciones de usuario de python

## Creación de listas

### Secuencia de creación de una lista de datos  

1. Cree una variable con el nombe del tipo de datos que albergará la lista 
2. Recuerde que la lista es una estructura delimitada por corchetes y cuyos elementos se separan por comas. 

In [1]:
edades = [14, 48, 73]
estaturas = [1.66, 1.65, 1.60]

In [3]:
type(edades), type(estaturas)

(list, list)

In [6]:
type(edades[0]), type(estaturas[0])

(int, float)

In [13]:
edades = [19, 20, 21]

## Acceso a los datos de una lista

In [14]:
edades[2] # se accede a los datos en una lista a través de la notación de corchetes
# y cada uno de los elementos de la lista tiene asociado un indice para cada posición. 

21

In [8]:
estaturas[2]

1.6

## creación de tuplas 

In [9]:
edades_tupla = (14, 48, 73)
estaturas_tuplas = (1.66, 1.65, 1.60) 

## Acceso a los datos de una tupla de Python 

In [11]:
edades[0]

14

In [10]:
estaturas[2] 

1.6

## Definición de media o promedio de una cantidad de datos cuantitatuva discreta o continua. 

$$ \bar{x} = \frac{\sum x_{i}}{n} $$

## Funciones incorporadas 

In [12]:
edades 

[14, 48, 73]

In [13]:
sum(edades)   

135

In [14]:
len(edades)

3

In [16]:
media = sum(edades)/len(edades) 
media 

45.0

In [14]:
tupla2 = (1,2,3) 

In [15]:
sum(tupla2) 

6

In [12]:
len(estaturas) 

3

## Funciones definidas por el usuario

In [17]:
def calcula_media(datos):
    suma = sum(datos)
    longitud = len(datos)
    return suma/longitud

In [18]:
calcula_media(edades)

45.0

In [1]:
import numpy as np 

In [17]:
edades_array = np.array(edades) 

In [18]:
type(edades_array)

numpy.ndarray

In [19]:
np.mean(edades_array)

20.0

# Calcular la media con scipy

In [23]:
import scipy
import numpy

In [24]:
numpy.mean(edades_array) 

20.0

# Calculo de la media con el módulo statistics

In [33]:
import statistics

In [35]:
type(statistics.mean(edades_array))

numpy.int32

1. Mean 

* Cómo se define
* Ejemplos ilustrativos.
* Visualización de esta medida de tendencia central 
* Cómo se interpreta. 

In [20]:
def calcule_media(numeros):
    return sum(numeros)/len(numeros)

In [21]:
calcule_media(edades) 

20.0

In [22]:
calcule_media(estaturas) 

1.7

Creación de una función de media personalizada

Comencemos calculando la edad promedio (media) de los jugadores de un equipo de baloncesto. El nombre del equipo será "Pythonic Machines".


In [12]:

team_ages = [19, 20, 21]


In [10]:
def promedio_de_edades(lista_edades):
    return sum(lista_edades)/len(lista_edades) 

In [13]:
promedio_de_edades(team_ages) 

20.0

In [6]:

def mean(dataset):
    return sum(dataset) / len(dataset)

print(mean(pythonic_machine_ages))

26.375


Desglosando este código:

* "ages" es una lista con las edades de los jugadores de baloncesto
* Definimos una función `media()` que devuelve la suma del conjunto de datos dado dividido por su longitud
* Los `sum()` La función devuelve la suma total (irónicamente) de los valores de un iterable, en este caso, una lista. Intente pasar el conjunto de datos como argumento, devolverá 211
* Los `len()` La función devuelve la longitud de un iterable, si le pasa el conjunto de datos obtendrá 8
* Pasamos las edades del equipo de baloncesto al `media()` función e imprimir el resultado.

Si verifica la salida, obtendrá:

## Tareas de fijación de los conceptos vistos y de adquisición de habilidades. 

1. Importe el dataset de florez de iris y determine el promedio o media por atributos. Y realizar las interpretacionesde estos estadísticos. 
2. Luego determine la media o promedio de los cuatro atributos para cada especie de flor de Iris. 

## Solución de la tarea de fijación y adquisición de habilidades

In [36]:
from sklearn.datasets import load_iris 

In [37]:
iris = load_iris(as_frame = False) 

In [38]:
iris.keys() 

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [39]:
X = iris.data

In [40]:
y = iris.target

In [41]:
X.shape

(150, 4)

In [45]:
len(np.mean(X, axis = 0)) # cacule media por columnas

4

Interpretaciones

1. la media de las longitudes de sepalo en este dataset de florez de iris es 5.84.
2. la media de las anchos de sepalo de las flores en este dataset es de 3.057.

## Evaluación 

1. Importe el dataset housing del capítulo 2 de Geron y determine la media de cada una de las variables cuantitativas continuas presentes en este dataset.  

## solución de la evaluación

In [1]:
import pandas as pd 

In [3]:
pd.read_csv('housing.csv').describe() 

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,20640.000000,20640.000000,20640.000000,20640.000000,20433.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,-119.569704,35.631861,28.639486,2635.763081,537.870553,1425.476744,499.539680,3.870671,206855.816909
std,2.003532,2.135952,12.585558,2181.615252,421.385070,1132.462122,382.329753,1.899822,115395.615874
min,-124.350000,32.540000,1.000000,2.000000,1.000000,3.000000,1.000000,0.499900,14999.000000
25%,-121.800000,33.930000,18.000000,1447.750000,296.000000,787.000000,280.000000,2.563400,119600.000000
50%,-118.490000,34.260000,29.000000,2127.000000,435.000000,1166.000000,409.000000,3.534800,179700.000000
75%,-118.010000,37.710000,37.000000,3148.000000,647.000000,1725.000000,605.000000,4.743250,264725.000000
max,-114.310000,41.950000,52.000000,39320.000000,6445.000000,35682.000000,6082.000000,15.000100,500001.000000


La edad promedio de las casas en los 20640 distritos es de $28.63$

In [4]:
from pathlib import Path
import pandas as pd
import tarfile
import urllib.request

def load_housing_data():
    tarball_path = Path("datasets/housing.tgz")
    if not tarball_path.is_file():
        Path("datasets").mkdir(parents=True, exist_ok=True)
        url = "https://github.com/ageron/data/raw/main/housing.tgz"
        urllib.request.urlretrieve(url, tarball_path)
        with tarfile.open(tarball_path) as housing_tarball:
            housing_tarball.extractall(path="datasets")
    return pd.read_csv(Path("datasets/housing/housing.csv"))

housing = load_housing_data()

In [6]:
X = housing.values

In [8]:
X[:,2].mean() 

28.639486434108527

## Tareas para realizar en casa

## Tareas para realizar durante la clase 

Realiza las actividades (debates, exposiciones orales, informes,
diario de aprendizaje, cuaderno de trabajo cooperativo,...).
Recibe retroalimentación inmediata del/la docente y de sus
compañeros y compañeras.
Accede, si fuera necesario, a las actividades previas facilitadas
por el/la docente.

* Entre la media y la madiana, cuál consideras como mejor medida de tenmdencia central y por qué?

## Evaluación

## Reflexiones y autoevaluación derivada de la realización de las tareas y la evaluación propuesta. 

Los datos generalmente vienen contenidos en una estructura de datos llamada nbarray de NumPy, es decir, vienen como grandes matrices. Y es por esto que el parátro axis de la función np.mean() cobra mucho valor. 

## Referentes 

* Medidas de tendencia central con python: https://geekflare.com/es/python-find-mean-median-and-mode/

* documentación de la función scipy.stats.mean(): https://www.geeksforgeeks.org/scipy-stats-mean-function-python/

* sobre la obtención del dataset de housing: https://github.com/ageron/handson-ml3/blob/main/02_end_to_end_machine_learning_project.ipynb

### [Continue la secuencia de enseñanza a través de este vínculo](https://github.com/marco-canas/didactica_ciencia_datos/blob/main/propuesta/unidad_didac/1_statistic/1_Basic_Descriptive_Statistics/1_medidas_tendencia_central/2_mediana.ipynb)